video/gazebocamerainput.h

```diff
@@ -0,0 +1,144 @@
+namespace Video {
+
+constexpr const char *GazeboCameraDeviceName = "Gazebo Camera";
+
+//----------------------------------------------------------------------------
+// BoBRobotics::Video::GazeboCameraInput
+//----------------------------------------------------------------------------
+//! A thin wrapper for reading from any video source supported by OpenCV
+class GazeboCameraInput : public Input
+{
+public:
+    //! Create a new video stream
+    GazeboCameraInput()
+      : GazeboCameraInput("/gazebo/default/my_simple_cart/camera/link/camera/image")
+    {}
+
+    /*!
+     * \brief Create a video stream for a specific device
+     *
+     * @param device Integer or string representation of device (passed to
+     *        cv::VideoCapture's constructor)
+     * @param cameraName The short name to use for this camera (see getCameraName())
+     */
+    
+    GazeboCameraInput(const std::string &topic, const std::string &cameraName = GazeboCameraDeviceName)
+      : m_CameraTopic(topic), m_CameraName(cameraName)
+    {
+        // Load gazebo as a client
+        #if GAZEBO_MAJOR_VERSION < 6
+        gazebo::setupClient(0, 0);
+        #else
+        gazebo::client::setup(0, 0);
+        #endif
+        //Setup image subscription node
+        m_ImageNode= gazebo::transport::NodePtr(new gazebo::transport::Node());
+        m_ImageNode->Init();
+        // Subscribe to the topic, and register a callback
+        m_ImageSub = m_ImageNode->Subscribe(m_CameraTopic, &GazeboCameraInput::OnImageMsg, this);
+        // m_OutSize.width = 320;
+        // m_OutSize.height = 240;
+
+        std::cerr << "Subsribed to "<< m_CameraTopic <<"\n";
+
+    }
+
+    /*!
+     * \brief Create a video stream for a specific device and a specified resolution
+     *
+     * @param device Integer or string representation of device (passed to
+     *        cv::VideoCapture's constructor)
+     * @param outSize Output resolution of camera
+     * @param cameraName The short name to use for this camera (see getCameraName())
+     */
+    GazeboCameraInput(const std::string &topic, const cv::Size &outSize,
+                const std::string &cameraName = GazeboCameraDeviceName)
+      : GazeboCameraInput(topic, cameraName)
+    {
+        setOutputSize(outSize);
+    }
+
+    //------------------------------------------------------------------------
+    // Video::Input virtuals
+    //------------------------------------------------------------------------
+    virtual std::string getCameraName() const override
+    {
+        return m_CameraName;
+    }
+
+    virtual cv::Size getOutputSize() const override
+    {
+        return m_OutSize;
+    }
+
+    virtual bool readFrame(cv::Mat &outFrame) override
+    {
+        outFrame = m_ReceivedImage;
+
+        // If there's no error, then we have updated frame and so return true
+        return true;
+    }
+
+    virtual void setOutputSize(const cv::Size &outSize) override
+    {
+        
+    }
```

Comment:
I assume it should be possible to change the resolution via a plugin but I haven't looked at it. I'll change the API for now and try to implement the programmatic change of resolution as soon as possible. It's currently being set in the model.sdf file for the differential_drive_robot in simulated_tank_gazebo.

Reply:
In that case, just throw an error here for now.

Reply:
I've removed this function entirely since the error is already being thrown in the superclass definition
```cpp
virtual void setOutputSize(const cv::Size &)
{
    throw std::runtime_error("This camera's resolution cannot be changed at runtime");
}
```
Is this alright?

Reply:
Cool. Yes, that's actually the right way to do this.